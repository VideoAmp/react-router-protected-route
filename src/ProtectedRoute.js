import React from "react";
import PropTypes from "prop-types";
import { Route, Redirect } from "react-router-dom";

const ProtectedRoute = ({
    isAccessible,
    redirectToPath,
    component: Component,
    ...rest
}) => (
    <Route
        {...rest}
        render={props => (
            isAccessible ?
                <Component {...props} /> :
                <Redirect
                    to={{
                        pathname: redirectToPath,
                        state: { from: props.location },
                    }}
                />
        )}
    />
);

ProtectedRoute.propTypes = {
    isAccessible: PropTypes.bool.isRequired,
    redirectToPath: PropTypes.string.isRequired,
    component: PropTypes.func.isRequired,
    location: PropTypes.shape({
        /** Current exact location of page */
        pathname: PropTypes.string,
    }),
};

ProtectedRoute.defaultProps = {
    location: {},
};

export default ProtectedRoute;
