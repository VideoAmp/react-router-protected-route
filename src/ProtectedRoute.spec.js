import React from "react";
import { mount } from "enzyme";
import { MemoryRouter, Route, Switch } from "react-router-dom";

import ProtectedRoute from "./ProtectedRoute";

describe("ProtectedRoute", () => {
    describe("is accessible", () => {
        it("renders component", () => {
            const tree = mount(
                <MemoryRouter
                    initialEntries={["/"]}
                    initialIndex={1}
                    keyLength={null}
                >
                    <Switch>
                        <ProtectedRoute
                            isAccessible
                            redirectToPath="/login"
                            path="/"
                            component={() => <p>Logged in</p>}
                        />
                        <Route
                            path="/login"
                            render={() => <p>Logged out</p>}
                        />
                    </Switch>
                </MemoryRouter>,
            );
            expect(tree).toMatchSnapshot();
        });
    });

    describe("is not accessible", () => {
        it("calls for redirect", () => {
            const tree = mount(
                <MemoryRouter
                    initialEntries={["/"]}
                    initialIndex={1}
                    keyLength={null}
                >
                    <Switch>
                        <ProtectedRoute
                            isAccessible={false}
                            redirectToPath="/login"
                            path="/"
                            component={() => <p>Logged in</p>}
                        />
                        <Route
                            path="/login"
                            render={() => <p>Logged out</p>}
                        />
                    </Switch>
                </MemoryRouter>,
            );
            expect(tree).toMatchSnapshot();
        });
    });
});
