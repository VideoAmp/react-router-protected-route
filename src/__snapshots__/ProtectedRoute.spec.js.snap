// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtectedRoute is accessible renders component 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/",
    ]
  }
  initialIndex={1}
  keyLength={null}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Switch>
      <ProtectedRoute
        component={[Function]}
        computedMatch={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        isAccessible={true}
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        path="/"
        redirectToPath="/login"
      >
        <Route
          computedMatch={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          path="/"
          render={[Function]}
        >
          <component
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "canGo": [Function],
                "createHref": [Function],
                "entries": Array [
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                ],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "index": 0,
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            staticContext={undefined}
          >
            <p>
              Logged in
            </p>
          </component>
        </Route>
      </ProtectedRoute>
    </Switch>
  </Router>
</MemoryRouter>
`;

exports[`ProtectedRoute is not accessible calls for redirect 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/",
    ]
  }
  initialIndex={1}
  keyLength={null}
>
  <Router
    history={
      Object {
        "action": "REPLACE",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/login",
            "search": "",
            "state": Object {
              "from": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
            },
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/login",
          "search": "",
          "state": Object {
            "from": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          },
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Switch>
      <ProtectedRoute
        component={[Function]}
        computedMatch={
          Object {
            "isExact": false,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        isAccessible={false}
        location={
          Object {
            "hash": "",
            "pathname": "/login",
            "search": "",
            "state": Object {
              "from": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
            },
          }
        }
        path="/"
        redirectToPath="/login"
      >
        <Route
          computedMatch={
            Object {
              "isExact": false,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/login",
              "search": "",
              "state": Object {
                "from": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              },
            }
          }
          path="/"
          render={[Function]}
        >
          <Redirect
            push={false}
            to={
              Object {
                "pathname": "/login",
                "state": Object {
                  "from": Object {
                    "hash": "",
                    "pathname": "/login",
                    "search": "",
                    "state": Object {
                      "from": Object {
                        "hash": "",
                        "pathname": "/",
                        "search": "",
                        "state": undefined,
                      },
                    },
                  },
                },
              }
            }
          />
        </Route>
      </ProtectedRoute>
    </Switch>
  </Router>
</MemoryRouter>
`;
